stake-pool/program/src/big_vec.rs

```diff
@@ -0,0 +1,245 @@
+            current_index: VEC_SIZE_BYTES,
+            inner: self,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Find matching data in the array
+    pub fn find<'b, T: Pack>(
+        &'a self,
+        data: &'b [u8],
+        predicate: fn(&[u8], &[u8]) -> bool,
+    ) -> Option<&T> {
+        let len = self.len() as usize;
+        let mut current = 0;
+        let mut current_index = VEC_SIZE_BYTES;
+        while current != len {
+            let end_index = current_index + T::LEN;
+            let current_slice = &self.data[current_index..end_index];
+            if predicate(current_slice, data) {
+                return Some(unsafe { &*(current_slice.as_ptr() as *const T) });
+            }
+            current_index = end_index;
+            current += 1;
+        }
+        None
+    }
+
+    /// Find matching data in the array
+    pub fn find_mut<'b, T: Pack>(
+        &'a mut self,
+        data: &'b [u8],
+        predicate: fn(&[u8], &[u8]) -> bool,
+    ) -> Option<&mut T> {
+        let len = self.len() as usize;
+        let mut current = 0;
+        let mut current_index = VEC_SIZE_BYTES;
+        while current != len {
+            let end_index = current_index + T::LEN;
+            let current_slice = &self.data[current_index..end_index];
+            if predicate(current_slice, data) {
+                return Some(unsafe { &mut *(current_slice.as_ptr() as *mut T) });
+            }
+            current_index = end_index;
+            current += 1;
+        }
+        None
+    }
+}
+
+/// Iterator wrapper over a BigVec
+pub struct Iter<'a, T> {
+    len: usize,
+    current: usize,
+    current_index: usize,
+    inner: &'a BigVec<'a>,
+    phantom: PhantomData<T>,
+}
+
+impl<'a, T: Pack + 'a> Iterator for Iter<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current == self.len {
+            None
+        } else {
+            let end_index = self.current_index + T::LEN;
+            let value = Some(unsafe {
+                &*(self.inner.data[self.current_index..end_index].as_ptr() as *const T)
+            });
+            self.current += 1;
+            self.current_index = end_index;
+            value
+        }
+    }
+}
+
+/// Iterator wrapper over a BigVec
+pub struct IterMut<'a, T> {
+    len: usize,
+    current: usize,
+    current_index: usize,
+    inner: &'a mut BigVec<'a>,
+    phantom: PhantomData<T>,
+}
+
+impl<'a, T: Pack + 'a> Iterator for IterMut<'a, T> {
+    type Item = &'a mut T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current == self.len {
+            None
+        } else {
+            let end_index = self.current_index + T::LEN;
+            let value = Some(unsafe {
+                &mut *(self.inner.data[self.current_index..end_index].as_ptr() as *mut T)
+            });
+            self.current += 1;
+            self.current_index = end_index;
+            value
+        }
+    }
+}
```

Comment:
Mildly terrifying that there's no tests for the partial vec reimplementation.  Can we lift some from the Rust `vec` test suite so we don't need to reinvent them?

Reply:
Yep, on it, looks like we have 2.7k lines of tests to work with :grin: